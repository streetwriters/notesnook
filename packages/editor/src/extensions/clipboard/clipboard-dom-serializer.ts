/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { DOMSerializer } from "@tiptap/pm/model";
import { Fragment, Schema } from "prosemirror-model";

export class ClipboardDOMSerializer extends DOMSerializer {
  static fromSchema(schema: Schema): ClipboardDOMSerializer {
    return (
      schema.cached.clipboardDomSerializer ||
      (schema.cached.clipboardDomSerializer = new ClipboardDOMSerializer(
        this.nodesFromSchema(schema),
        this.marksFromSchema(schema)
      ))
    );
  }

  serializeFragment(
    fragment: Fragment,
    options?: { document?: Document | undefined } | undefined,
    target?: HTMLElement | DocumentFragment | undefined
  ): HTMLElement | DocumentFragment {
    const dom = super.serializeFragment(fragment, options, target);
    for (const p of dom.querySelectorAll("li > p")) {
      if (p.parentElement && p.parentElement.childElementCount > 1) continue;
      p.parentElement?.append(...p.childNodes);
      p.remove();
    }

    for (const element of dom.querySelectorAll("[data-block-id]")) {
      element.removeAttribute("data-block-id");
    }

    for (const p of dom.querySelectorAll('p[data-spacing="single"]')) {
      if (!p.previousElementSibling || p.previousElementSibling.tagName !== "P")
        continue;
      if (p.previousElementSibling.childNodes.length > 0)
        p.previousElementSibling.appendChild(document.createElement("br"));
      p.previousElementSibling.append(...p.childNodes);
      p.remove();
    }

    return dom;
  }
}
