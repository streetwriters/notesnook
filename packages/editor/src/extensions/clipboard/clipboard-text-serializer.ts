/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { TextSerializer } from "@tiptap/core";
import { Schema, Slice } from "prosemirror-model";
import { ListItem } from "../list-item/index.js";
import { EditorView } from "@tiptap/pm/view";

export function clipboardTextSerializer(content: Slice, view: EditorView) {
  return getTextBetween(content, view.state.schema);
}

export function getTextBetween(slice: Slice, schema: Schema): string {
  const range = { from: 0, to: slice.size };
  const separator = "\n";
  let text = "";
  let separated = true;

  slice.content.nodesBetween(0, slice.size, (node, pos, parent, index) => {
    const textSerializer = schema.nodes[node.type.name]?.spec
      .toText as TextSerializer;

    if (textSerializer) {
      if (node.isBlock && !separated) {
        text += separator;
        separated = true;
      }

      if (parent) {
        text += textSerializer({
          node,
          pos,
          parent,
          index,
          range
        });
      }
    } else if (node.isText) {
      text += node?.text;
      separated = false;
    } else if (node.isBlock && !!text) {
      // we don't want double spaced list items when pasting
      if (index === 0 && parent?.type.name === ListItem.name) return;

      text += separator;
      if (node.attrs.spacing === "double" && node.childCount > 0)
        text += separator;
      separated = true;
    }
  });

  return text;
}
