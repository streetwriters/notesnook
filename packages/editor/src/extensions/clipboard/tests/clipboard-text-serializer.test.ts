/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { test } from "vitest";
import { createEditor, h } from "../../../../test-utils/index.js";
import OrderedList from "../../ordered-list/index.js";
import { ListItem } from "../../list-item/index.js";
import { transformCopied } from "../index.js";
import { Paragraph } from "../../paragraph/index.js";
import { ClipboardDOMSerializer } from "../clipboard-dom-serializer.js";
import { clipboardTextSerializer } from "../clipboard-text-serializer.js";
import Link from "../../link/index.js";

test("copied list items shouldn't contain extra newlines", (t) => {
  const { editor } = createEditor({
    initialContent: h("div", [
      h("ol", [
        h("li", [
          h("p", ["This is line: ", h("em", ["number 1."])]),
          h("p", ["And this is line number 2."])
        ]),
        h("li", [h("p", ["This is line number 3."])])
      ])
    ]).innerHTML,
    extensions: {
      orderedList: OrderedList,
      listItem: ListItem
    }
  });

  const serializer = ClipboardDOMSerializer.fromSchema(
    editor.view.state.schema
  );
  t.expect(
    new XMLSerializer().serializeToString(
      serializer.serializeFragment(
        editor.state.doc.slice(0, editor.state.doc.nodeSize - 2).content
      )
    )
  ).toBe(
    "<ol><li><p>This is line: <em>number 1.</em></p><p>And this is line number 2.</p></li><li>This is line number 3.</li></ol>"
  );

  t.expect(
    clipboardTextSerializer(
      editor.state.doc.slice(0, editor.state.doc.nodeSize - 2),
      editor.view
    )
  ).toBe(`This is line: number 1.
And this is line number 2.
This is line number 3.`);
});

test("copying a single list item shouldn't copy the list metadata", (t) => {
  const { editor } = createEditor({
    initialContent: h("div", [h("ol", [h("li", ["Hello"])])]).innerHTML,
    extensions: {
      orderedList: OrderedList,
      listItem: ListItem
    }
  });

  t.expect(
    transformCopied(
      editor.state.doc.slice(0, editor.state.doc.nodeSize - 2),
      editor.view
    ).toJSON()
  ).toMatchSnapshot();
});

test("copying text from a list item shouldn't add extra spaces at the end", (t) => {
  const { editor } = createEditor({
    initialContent: h("div", [
      h("ol", [
        h("li", [
          h("p", ["I am ", h("a", ["Hello"], { href: "https://google.com/" })])
        ])
      ])
    ]).innerHTML,
    extensions: {
      orderedList: OrderedList,
      listItem: ListItem,
      link: Link
    }
  });

  t.expect(
    transformCopied(
      editor.state.doc.slice(
        editor.state.doc.nodeSize - 10,
        editor.state.doc.nodeSize - 2
      ),
      editor.view
    ).toJSON()
  ).toMatchSnapshot();
});

test("copying multiple lists shouldn't copy only the first list", (t) => {
  const { editor } = createEditor({
    initialContent: h("div", [
      h("ol", [h("li", ["Hello"])]),
      h("ol", [h("li", ["Hello 2"])]),
      h("ol", [h("li", ["Hello 3"])])
    ]).innerHTML,
    extensions: {
      orderedList: OrderedList,
      listItem: ListItem
    }
  });

  t.expect(
    transformCopied(
      editor.state.doc.slice(0, editor.state.doc.nodeSize - 2),
      editor.view
    ).toJSON()
  ).toMatchSnapshot();
});

test("copying a single nested list item shouldn't copy the list metadata", (t) => {
  const { editor } = createEditor({
    initialContent: h("div", [
      h("ol", [
        h("li", [
          h("p", ["Hello"]),
          h("ol", [h("li", [h("p", ["Nested list item"])])])
        ])
      ])
    ]).innerHTML,
    extensions: {
      orderedList: OrderedList,
      listItem: ListItem
    }
  });

  t.expect(
    transformCopied(
      editor.state.doc.slice(12, editor.state.doc.nodeSize - 2),
      editor.view
    ).toJSON()
  ).toMatchSnapshot();
});

const createParagraphs = (spacing: "double" | "single") => [
  h("p", ["I am paragraph 1."], { "data-spacing": spacing }),
  h("p", ["I am paragraph 2."], { "data-spacing": spacing }),
  h("p", ["I am paragraph 3."], { "data-spacing": spacing })
];

const createParagraphsWithSpaces = (spacing: "double" | "single") => [
  h("p", ["I am paragraph 1."], { "data-spacing": spacing }),
  h("p", [h("br")], { "data-spacing": spacing }),
  h("p", [h("br")], { "data-spacing": spacing }),
  h("p", ["I am paragraph 2."], { "data-spacing": spacing }),
  h("p", [h("br")], { "data-spacing": spacing }),
  h("p", [h("br")], { "data-spacing": spacing }),
  h("p", ["I am paragraph 3."], { "data-spacing": spacing })
];

const paragraphTestCases = [
  {
    spacing: "double",
    content: createParagraphs("double"),
    expectedHtml: `<p data-spacing="double">I am paragraph 1.</p><p data-spacing="double">I am paragraph 2.</p><p data-spacing="double">I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\n\nI am paragraph 2.\n\nI am paragraph 3.`
  },
  {
    spacing: "single",
    content: createParagraphs("single"),
    expectedHtml: `<p data-spacing="single">I am paragraph 1.<br>I am paragraph 2.<br>I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\nI am paragraph 2.\nI am paragraph 3.`
  },
  {
    spacing: "double",
    content: createParagraphsWithSpaces("double"),
    expectedHtml: `<p data-spacing="double">I am paragraph 1.</p><p data-spacing="double"></p><p data-spacing="double"></p><p data-spacing="double">I am paragraph 2.</p><p data-spacing="double"></p><p data-spacing="double"></p><p data-spacing="double">I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\n\n\n\nI am paragraph 2.\n\n\n\nI am paragraph 3.`
  },
  {
    spacing: "single",
    content: createParagraphsWithSpaces("single"),
    expectedHtml: `<p data-spacing="single">I am paragraph 1.<br><br><br>I am paragraph 2.<br><br><br>I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\n\n\nI am paragraph 2.\n\n\nI am paragraph 3.`
  },
  {
    spacing: "mixed",
    content: [
      ...createParagraphs("double"),
      ...createParagraphs("single"),
      ...createParagraphs("double")
    ],
    expectedHtml: `<p data-spacing="double">I am paragraph 1.</p><p data-spacing="double">I am paragraph 2.</p><p data-spacing="double">I am paragraph 3.<br>I am paragraph 1.<br>I am paragraph 2.<br>I am paragraph 3.</p><p data-spacing="double">I am paragraph 1.</p><p data-spacing="double">I am paragraph 2.</p><p data-spacing="double">I am paragraph 3.</p>`,
    expectedText: `I am paragraph 1.\n\nI am paragraph 2.\n\nI am paragraph 3.\nI am paragraph 1.\nI am paragraph 2.\nI am paragraph 3.\n\nI am paragraph 1.\n\nI am paragraph 2.\n\nI am paragraph 3.`
  },
  {
    spacing: "mixed 2",
    content: [
      h("p", ["I am paragraph 1."], { "data-spacing": "double" }),
      h("p", ["I am paragraph 2."], { "data-spacing": "single" }),
      h("p", ["I am paragraph 3."], { "data-spacing": "double" }),
      h("p", ["I am paragraph 4."], { "data-spacing": "single" }),
      h("p", ["I am paragraph 5."], { "data-spacing": "single" }),
      h("p", ["I am paragraph 6."], { "data-spacing": "double" })
    ],
    expectedHtml: `<p data-spacing="double">I am paragraph 1.<br>I am paragraph 2.</p><p data-spacing="double">I am paragraph 3.<br>I am paragraph 4.<br>I am paragraph 5.</p><p data-spacing="double">I am paragraph 6.</p>`,
    expectedText: `I am paragraph 1.\nI am paragraph 2.\n\nI am paragraph 3.\nI am paragraph 4.\nI am paragraph 5.\n\nI am paragraph 6.`
  }
];
for (const testCase of paragraphTestCases) {
  test(`copying should respect paragraph spacing (${testCase.spacing})`, (t) => {
    const { editor } = createEditor({
      initialContent: h("div", testCase.content).innerHTML,
      extensions: {
        paragraph: Paragraph,
        hardBreak: false
      }
    });

    const serializer = ClipboardDOMSerializer.fromSchema(
      editor.view.state.schema
    );
    t.expect(
      new XMLSerializer().serializeToString(
        serializer.serializeFragment(
          editor.state.doc.slice(0, editor.state.doc.nodeSize - 2).content
        )
      )
    ).toBe(testCase.expectedHtml);

    t.expect(
      clipboardTextSerializer(
        editor.state.doc.slice(0, editor.state.doc.nodeSize - 2),
        editor.view
      )
    ).toBe(testCase.expectedText);
  });
}
