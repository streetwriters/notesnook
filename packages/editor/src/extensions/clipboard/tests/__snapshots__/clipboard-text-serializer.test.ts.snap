// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`copying a single list item shouldn't copy the list metadata 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "openEnd": 1,
  "openStart": 1,
}
`;

exports[`copying a single nested list item shouldn't copy the list metadata 1`] = `
{
  "content": [
    {
      "content": [
        {
          "text": "Nested list item",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "openEnd": 1,
  "openStart": 1,
}
`;

exports[`copying multiple lists shouldn't copy only the first list 1`] = `
{
  "content": [
    {
      "attrs": {
        "listType": "",
        "start": 1,
        "type": undefined,
      },
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "text": "Hello",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
      ],
      "type": "orderedList",
    },
    {
      "attrs": {
        "listType": "",
        "start": 1,
        "type": undefined,
      },
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "text": "Hello 2",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
      ],
      "type": "orderedList",
    },
    {
      "attrs": {
        "listType": "",
        "start": 1,
        "type": undefined,
      },
      "content": [
        {
          "content": [
            {
              "content": [
                {
                  "text": "Hello 3",
                  "type": "text",
                },
              ],
              "type": "paragraph",
            },
          ],
          "type": "listItem",
        },
      ],
      "type": "orderedList",
    },
  ],
}
`;

exports[`copying text from a list item shouldn't add extra spaces at the end 1`] = `
{
  "content": [
    {
      "content": [
        {
          "marks": [
            {
              "attrs": {
                "class": null,
                "href": "https://google.com/",
                "rel": "noopener noreferrer nofollow",
                "target": "_blank",
                "title": null,
              },
              "type": "link",
            },
          ],
          "text": "Hello",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "openEnd": 1,
  "openStart": 1,
}
`;
